Site/SiteAnalyticsModule.php

```diff
@@ -221,9 +254,120 @@ public function getGoogleAnalyticsTrackerInlineJavascript()
 	}
 
 	// }}}
-	// {{{ protected function getTrackingCodeSource()
+	// {{{ public function hasFacebookPixel()
+
+	public function hasFacebookPixel()
+	{
+		return (
+			$this->facebook_pixel_id != '' &&
+			!$this->analytics_opt_out
+		);
+	}
+
+	// }}}
+	// {{{ public function pushFacebookPixelCommands()
+
+	public function pushFacebookPixelCommands(array $commands)
+	{
+		foreach ($commands as $command) {
+			$this->facebook_pixel_commands[] = $command;
+		}
+	}
+
+	// }}}
+	// {{{ public function prependFacebookPixelCommands()
+
+	public function prependFacebookPixelCommands(array $commands)
+	{
+		foreach ($commands as $command) {
+			array_unshift($this->facebook_pixel_commands, $command);
+		}
+	}
+
+	// }}}
+	// {{{ public function getFacebookPixelImage()
+
+	public function getFacebookPixelImage()
+	{
+		$xhtml = <<<'XHTML'
+<noscript><img height="1" width="1" style="display:none" src="https://www.facebook.com/tr?id=%s&ev=PageView&noscript=1"/></noscript>
+XHTML;
+
+		return sprintf(
+			$xhtml,
+			SwatString::minimizeEntities($this->facebook_pixel_id)
+		);
+	}
+
+	// }}}
+	// {{{ public function getFacebookPixelInlineJavascript()
+
+	public function getFacebookPixelInlineJavascript()
+	{
+		$javascript = null;
+
+		if ($this->hasFacebookPixel() &&
+			count($this->facebook_pixel_commands) > 0) {
+			$javascript = $this->getFacebookPixelTrackerInlineJavascript();
+			$javascript.= "\n";
+			$javascript.= $this->getFacebookPixelCommandsInlineJavascript();
+		}
+
+		return $javascript;
+	}
+
+	// }}}
+	// {{{ public function getFacebookPixelTrackerInlineJavascript()
 
-	protected function getTrackingCodeSource()
+	public function getFacebookPixelTrackerInlineJavascript()
+	{
+		$javascript = null;
+
+		if ($this->hasFacebookPixel()) {
+			$javascript = <<<'JS'
+!function(f,b,e,v,n,t,s){if(f.fbq)return;n=f.fbq=function(){n.callMethod?
+n.callMethod.apply(n,arguments):n.queue.push(arguments)};if(!f._fbq)f._fbq=n;
+n.push=n;n.loaded=!0;n.version='2.0';n.queue=[];t=b.createElement(e);t.async=!0;
+t.src=v;s=b.getElementsByTagName(e)[0];s.parentNode.insertBefore(t,s)}(window,
+document,'script','//connect.facebook.net/en_US/fbevents.js');
+JS;
+		}
+
+		return $javascript;
+	}
+
+	// }}}
+	// {{{ public function getFacebookPixelCommandsInlineJavascript()
+
+	public function getFacebookPixelCommandsInlineJavascript()
+	{
+		$javascript = null;
+
+		if ($this->hasFacebookPixel() &&
+			count($this->facebook_pixel_commands) > 0) {
+			$javascript = '';
+
+			// Always init with the account and track the pageview before any
+			// further commands.
+			$javascript.= $this->getFacebookPixelCommand(
```

Comment:
Could just assign `$javascript` here. The previous statement sets it to an empty string.